# **What is the relationship between specific post characteristics and the likelihood of a post achieving viral status on Social Media?**

# **Abstract**

With viral posts frequently emerging on social media platforms, understanding the factors that contribute to their widespread popularity has become increasingly important for social media influencers or brand companies. This study explores the relationship between specific post characteristics and the likelihood of a post achieving viral status. By analysing features such as content attributes, format, and engagement metrics, we identify patterns associated with highly shared, liked or commented posts. The findings aim to provide insights into the mechanisms behind online virality and offer guidance for creating content with higher potential for widespread reach.

# **Table Of Contents**

1. Key Terms & Definitions (Ask)
2. Analysis Goal (Ask)
3. Data Requirements Analysis (Prepare)
4. Data Sources (Prepare)
5. Determing DS Suitability (Prepare)
6. Data Preparation (Process)
7. Data Analysis (Analyze)

# **1. Key Terms & Definitions (Ask)**

**Post characteristics**

Post characteristics refer to measurable attributes of a social media post that may influence its popularity. These characteristics vary depending on the content format. For video posts, relevant features may include video duration, frame rate, and resolution. For text-based posts, relevant features may include text length, sentiment and tone. Additional attributes applicable across formats include posting time (hour of the day), day of the week, number of hashtags and creator status, to name a few.

**Viral status**

Viral status refers to a post achieving significantly higher visibility and engagement compared to typical posts on the same platform. Viral posts often exhibit rapid growth in engagement metrics, such as views, likes, comments, and shares, indicating a fast spread across users. This spread may extend beyond the post’s original audience or community. In cases where temporal engagement data is unavailable (growth rate of views against time for e.g.), virality will be determined using predefined thresholds within a fixed time frame. This can include reaching a specific number of views, shares, or likes that is equivalent to the engagement median times three, or the post is in the top 5% by engagement in its dataset.

**Social Media**

This study explores posts from TikTok and Instagram. These platforms are among the most widely used social media platform globally and provide diverse content formats such as short videos, images, and text-based posts, making them suitable platforms for analysing patterns of online virality.

# **2. Analysis Goal (Ask)**

The primary objective of this analysis is to determine how specific characteristics of social media posts influence their likelihood of achieving viral status. The main question to be asked in this study is: Which post characteristics are most strongly associated with viral posts on social media platforms?

To address this research question, we first conduct exploratory data analysis to identify patterns in the characteristics of viral posts. Post characteristics such as content length, number of hashtags, posting time, and media properties are analysed to determine their relationship with virality. Using these insights, a predictive model is then developed to estimate the likelihood of a post becoming viral based on its attributes.

The performance and usefulness of the analysis will be evaluated based on its ability to identify meaningful patterns and predictive features associated with viral content. Depending on the modelling approach used, evaluation metrics may include classification accuracy, precision, recall, or other relevant performance indicators.

Stakeholders can then strategically incorporate such characteristics into their posts to increase the probability of making it viral.

# **3. Data Requirements Analysis (Prepare)**

To tackle the problem, several categories of data are required.

First, post metadata is needed to provide contextual information about when and where a post was published. This includes the date and time of posting, the day of the week, the account used to publish the post, and the number of followers the account had at the time of posting.

Second, engagement metrics are required to determine whether a post achieves viral status. These metrics include the number of likes or reactions, the number of shares, and the number of comments. Additional indicators such as reach or impressions, engagement rate (calculated as total engagement divided by the number of followers), and the growth rate of engagement within a specific timeframe may also be collected if available.

Third, content characteristics describe the properties of the post itself. These include the content type (such as text-only, image, or video), video length, text length, and the number of images included in a post. Textual features may also be analysed, including caption length, the number of hashtags, mentions or tags, the presence of emojis, sentiment score (e.g., positive, negative, or neutral), and readability score. Where available, visual features such as image colorfulness or brightness, the presence of faces or people, thumbnail quality for videos, and audio-related information such as sound effects may also be considered.

Finally, additional contextual information may be collected to further enrich the analysis. This includes the topic category of the post, which may help identify patterns related to content themes or audience characteristics.

# **4. Data Sources (Prepare)**

**List of considered data sources:**

In the following, we outline several potential data sources that were explored for this project. We evaluate each source for their suitability for analysing social media virality. The final selected data source is marked with an asterisk (*).

**Instagram DS**
1. DS1 - Link: [Instagram Analytics](https://www.kaggle.com/datasets/kundanbedmutha/instagram-analytics-dataset) - `Instagram_Analytics.csv` *
2. DS2 - Link: [Instagram Posts](https://brightdata.com/cp/datasets/browse/gd_lk5ns7kz21pck8jpis?id=hl_6c69a494) - `Instagram_Posts.csv` *
3. DS3 - Link: [Instagram Reels](https://brightdata.com/cp/datasets/browse/gd_lyclm20il4r5helnj?id=hl_6c69a494) - `Instagram_Reels.csv` *
4. DS4 - Link: [Instagram Datasets](https://github.com/luminati-io/Social-media-dataset-samples) - `Instagram_Datasets.csv`

# **UNSURE FOR TIKTOK FOR NOW BECAUSE ITS JUST VIDEOS AND UNABLE TO COMPARE THE CONTENT OF VIDEOS**
**TikTok DS**
1. DS5 - Link: [TikTok Posts](https://huggingface.co/datasets/Shofo/shofo-tiktok-general-small) - `TikTok_Posts.csv` *
2. DS6 - Link: [TikTok Texts](https://huggingface.co/datasets/Shofo/shofo-tiktok-general-small) - `TikTok_Texts.csv` *

Instagram datasets will serve as the primary training ground for the initial modeling pipeline. Once the architecture is validated, it will be applied to the TikTok data to explore cross-platform patterns and the robustness of the model.

These datasets were assessed based on their availability of engagement metrics, post metadata, and content features required for the analysis. The selected dataset (*) provides the most comprehensive set of variables needed to examine the relationship between post attributes and social media virality.

# **5. Determing DS Suitability (Prepare)**

In [1397]:
import pandas as pd

## **DS1 - Instagram Analytics**

In [1398]:
df_analytics = pd.read_csv("datasets/raw/Instagram_Analytics.csv")

# Structural profile
print(df_analytics.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29999 entries, 0 to 29998
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   post_id                   29999 non-null  object 
 1   account_id                29999 non-null  int64  
 2   account_type              29999 non-null  object 
 3   follower_count            29999 non-null  int64  
 4   media_type                29999 non-null  object 
 5   content_category          29999 non-null  object 
 6   traffic_source            29999 non-null  object 
 7   has_call_to_action        29999 non-null  int64  
 8   post_datetime             29999 non-null  object 
 9   post_date                 29999 non-null  object 
 10  post_hour                 29999 non-null  int64  
 11  day_of_week               29999 non-null  object 
 12  likes                     29999 non-null  int64  
 13  comments                  29999 non-null  int64  
 14  shares

`Instagram_Analytics.csv` is a structured post-level dataset containing content features (`media_type`, `content_category`, `hashtags_count`, `caption_length`, `post_hour`, `day_of_week`, `post_date`), engagement outcomes (`likes`, `comments`, `shares`, `saves`, `reach`, `impressions`, `engagement_rate`), and performance labels (`performance_bucket_label` and `viral` in sampled rows). This aligns strongly with our analytical task because it directly supports studying the relationship between post characteristics and viral outcomes.


**Exploring media_type and content_category and post_hour**

In [1399]:
media_type_counts = df_analytics.groupby('media_type')['media_type'].count().sort_values(ascending=False)
print(media_type_counts, '\n')

content_category_counts = df_analytics.groupby('content_category')['content_category'].count().sort_values(ascending=False)
print(content_category_counts, '\n')

hourly_post_counts = df_analytics.groupby('post_hour')['post_hour'].count().sort_values(ascending=False)
print('top 5 post hours')
print(hourly_post_counts.head(5))

media_type
image       11927
carousel    10627
reel         7445
Name: media_type, dtype: int64 

content_category
Photography    3035
Fashion        3034
Technology     3025
Lifestyle      3017
Food           3010
Fitness        3004
Music          3003
Travel         2968
Beauty         2953
Comedy         2950
Name: content_category, dtype: int64 

top 5 post hours
post_hour
11    1283
14    1282
22    1280
13    1280
0     1277
Name: post_hour, dtype: int64


The dataset shows a relatively balanced distribution across content categories, with each of the 10 niches maintaining a steady volume of approximately 3,000 posts. However, in terms of media type, static images and carousels are the primary drivers of volume, collectively accounting for nearly 75% of the total media shared, while reels represent the smallest portion of the mix. We can see that the post hours is also in 24 hours format.

These categories will be useful in identifying the recommended format and topic for a viral post.

In [1400]:
# Missingness and duplicates
print("\nMissing values:\n", df_analytics.isna().sum().sort_values(ascending=False))
print("\nDuplicate post_id:", df_analytics["post_id"].duplicated().sum())


Missing values:
 post_id                     0
account_id                  0
account_type                0
follower_count              0
media_type                  0
content_category            0
traffic_source              0
has_call_to_action          0
post_datetime               0
post_date                   0
post_hour                   0
day_of_week                 0
likes                       0
comments                    0
shares                      0
saves                       0
reach                       0
impressions                 0
engagement_rate             0
followers_gained            0
caption_length              0
hashtags_count              0
performance_bucket_label    0
dtype: int64

Duplicate post_id: 0


DS1 has no null entries identified. No duplicate records exist after verifying the `post_id` field for uniqueness. The clean dataset ensures a robust foundation for subsequent engagement analysis.

In [1401]:
# Value quality checks for numeric metrics
# Negative engagement values are invalid and should be treated as data quality issues.
print("Number of negative values:")
for col in ["follower_count", "likes", "comments", "shares", "saves", "reach", "impressions", "caption_length", "hashtags_count"]:
    print(col, ":", (df_analytics[col] < 0).sum())

Number of negative values:
follower_count : 0
likes : 0
comments : 0
shares : 0
saves : 0
reach : 0
impressions : 0
caption_length : 0
hashtags_count : 0


Check revealed no negative values, so there are no invalid rows.

In [1402]:
# Outcome and feature distribution checks
distribution_of_viral = df_analytics.groupby('performance_bucket_label')["performance_bucket_label"].count().sort_values(ascending=False)
print(distribution_of_viral, '\n')

print(df_analytics[["likes", "comments", "shares", "hashtags_count", "caption_length"]].describe())

performance_bucket_label
high      7500
low       7500
viral     7500
medium    7499
Name: performance_bucket_label, dtype: int64 

              likes      comments        shares  hashtags_count  \
count  29999.000000  29999.000000  29999.000000    29999.000000   
mean     287.653588      8.521917     14.426614        7.987933   
std      317.647682     10.116505     16.420899        2.828130   
min        0.000000      0.000000      0.000000        0.000000   
25%      104.000000      3.000000      5.000000        6.000000   
50%      199.000000      6.000000     10.000000        8.000000   
75%      363.000000     11.000000     19.000000       10.000000   
max    10632.000000    339.000000    516.000000       21.000000   

       caption_length  
count    29999.000000  
mean       120.081169  
std         11.009108  
min         70.000000  
25%        113.000000  
50%        120.000000  
75%        127.000000  
max        166.000000  


Each category for `performance_bucket_label` has exactly 7,500 entries (with `medium` missing just one). The dataset is balanced. The engagement metrics (`likes`,`comments`,`shares`) are heavily right-skewed and confirms the presence of outliers. The other metrics (`hashtags_count`,`caption_length`) are more normally distributed.

### Data Quality Discussion
- **Positive criterion (Relevance/Completeness):** The dataset contains both explanatory post-characteristic variables and engagement outcomes in one place, making it highly suitable for the main research question.
- **Problematic criterion (Potential label leakage):** `performance_bucket_label` may be derived from engagement metrics (`likes`, `comments`, `shares`, etc.). If so, using the same engagement metrics as predictors for that label can cause leakage.

### Concluding Suitability Statement
This dataset is **highly suitable** and should be a primary source for the analytical phase. Key precautions are to verify how `engagement_rate` and `performance_bucket_label` are defined and prevent label leakage by excluding features that directly define the target when building predictive models.

## **DS2 - Instagram Posts**

In [1403]:
df_posts = pd.read_csv("datasets/raw/Instagram_Posts.csv")

# Structural profile
print(df_posts.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   url                    1000 non-null   object 
 1   user_posted            1000 non-null   object 
 2   description            956 non-null    object 
 3   hashtags               647 non-null    object 
 4   num_comments           1000 non-null   int64  
 5   date_posted            1000 non-null   object 
 6   likes                  888 non-null    float64
 7   photos                 1000 non-null   object 
 8   videos                 179 non-null    object 
 9   location               276 non-null    object 
 10  latest_comments        583 non-null    object 
 11  post_id                1000 non-null   int64  
 12  discovery_input        0 non-null      float64
 13  has_handshake          0 non-null      float64
 14  shortcode              1000 non-null   object 
 15  conte

`Instagram_Posts.csv` is a post-level dataset with rich metadata (40 columns including `description`, `hashtags`, `date_posted`, `content_type`) and core engagement fields (`likes`, `num_comments`). It also aligns with our analytical task. However, no explicit viral label is provided, so we must derive virality ourselves from engagement thresholds.

**Exploring content_type and is_verified**

In [1404]:
content_type_counts = df_posts.groupby('content_type')['content_type'].count().sort_values(ascending=False)
print(content_type_counts, '\n')

verified_counts = df_posts.groupby('is_verified')['is_verified'].count().sort_values(ascending=False)
print(verified_counts, '\n')

content_type
Image       507
Carousel    371
Reel         80
Video        42
Name: content_type, dtype: int64 

is_verified
False    736
True     264
Name: is_verified, dtype: int64 



DS2 has similar content type as DS1. This dataset can be used to supplement DS1. We can consider reel and video to be of similar type for the future.

In [1405]:
# Missingness and duplicates
print("\nMissing values:\n", df_posts.isna().sum().sort_values(ascending=False))
print("Duplicate post_id:", df_posts["post_id"].duplicated().sum())


Missing values:
 discovery_input          1000
has_handshake            1000
audio_url                 983
video_play_count          894
coauthor_producers        893
engagement_score_view     878
product_type              878
video_view_count          834
videos                    821
videos_duration           769
tagged_users              730
location                  724
latest_comments           417
hashtags                  353
alt_text                  132
likes                     112
description                44
photos                      0
num_comments                0
url                         0
user_posted                 0
date_posted                 0
post_id                     0
shortcode                   0
content_id                  0
pk                          0
thumbnail                   0
content_type                0
is_verified                 0
followers                   0
posts_count                 0
profile_image_link          0
user_posted_id        

Data profiling revealed a substantial number of missing values in this dataset. First, features `discovery_input` and `has_handshake` contained 100% null entries, rendering them uninformative. Second, a high sparsity in video-specific metrics (e.g., `video_play_count` at 89.4% null) contrasted with the 100% completion of `photos_number`, confirming the dataset is predominantly static image content. Finally, the uniqueness of post_id confirms a high level of record integrity with no duplicates.

In [1406]:
# Value quality checks for numeric metrics
# Negative engagement values are invalid and should be treated as data quality issues.
print("Number of negative values:")
for col in ["likes", "num_comments", "followers"]:
    print(col, ":", (df_posts[col] < 0).sum())

Number of negative values:
likes : 0
num_comments : 0
followers : 0


Check revealed no negative values, so there are no invalid rows.

### Data Quality Discussion
- **Positive criterion (Feature richness):** Strong variety of post and account characteristics supports meaningful feature engineering (e.g., hashtag count, posting time, creator-level controls).
- **Problematic criterion (Completeness consistency):** Several fields are sparse or format-inconsistent across rows (e.g., nested text/metadata fields). This requires careful column selection and standardised cleaning before analysis.

### Concluding Suitability Statement
This dataset is **suitable with moderate preprocessing effort**. It is valuable for post-feature extraction and exploratory relationship testing, but requires robust cleaning and explicit virality definition before predictive analysis.

## **DS3 - Instagram Reels**

In [1407]:
df_reels = pd.read_csv("datasets/raw/Instagram_Reels.csv")

# Structural profile
print(df_reels.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   url                  1000 non-null   object 
 1   user_posted          1000 non-null   object 
 2   description          956 non-null    object 
 3   hashtags             656 non-null    object 
 4   num_comments         1000 non-null   int64  
 5   date_posted          1000 non-null   object 
 6   likes                1000 non-null   int64  
 7   views                1000 non-null   int64  
 8   video_play_count     984 non-null    float64
 9   top_comments         1000 non-null   object 
 10  post_id              1000 non-null   object 
 11  thumbnail            1000 non-null   object 
 12  shortcode            1000 non-null   object 
 13  content_id           1000 non-null   object 
 14  product_type         1000 non-null   object 
 15  coauthor_producers   1000 non-null   ob

`Instagram_Reels.csv` contains reel-level records with strong video-centric attributes (`views`, `video_play_count`, `length`, `audio_url`) plus engagement (`likes`, `num_comments`) and account context (`followers`, `following`). It is highly relevant for studying virality in short-form video content and complements static post data (DS2).

In [1408]:
# Missingness and duplicates
print("\nMissing values:\n", df_reels.isna().sum().sort_values(ascending=False))
print("Duplicate post_id:", df_reels["post_id"].duplicated().sum())


Missing values:
 following              1000
posts_count             481
followers               481
hashtags                344
audio_url                48
description              44
video_play_count         16
profile_image_link        1
user_posted               0
date_posted               0
url                       0
post_id                   0
top_comments              0
views                     0
likes                     0
num_comments              0
thumbnail                 0
content_id                0
shortcode                 0
video_url                 0
length                    0
tagged_users              0
coauthor_producers        0
product_type              0
user_profile_url          0
is_paid_partnership       0
is_verified               0
dtype: int64
Duplicate post_id: 0


Data profiling revealed a few missing values in this dataset. First, features `following` contained 100% null entries, rendering it uninformative. The account-level metrics `posts_count` and `followers` are each missing 48.1% of their data. The other columns with null values are relatively small in numbers and should be easy to handle. Finally, the uniqueness of post_id confirms a high level of record integrity with no duplicates.

In [1409]:
# Value quality checks for numeric metrics
# Negative engagement values are invalid and should be treated as data quality issues.
for col in ["num_comments", "likes", "views", "followers", "video_play_count", "length"]:
    print(col, ":", (df_reels[col] < 0).sum())

num_comments : 0
likes : 57
views : 0
followers : 0
video_play_count : 0
length : 0


We identified 57 records with negative values in the `likes` field. Given that Instagram likes are strictly non-negative integers, and that Instagram does not have any dislike feature, these records will be flagged as data corruption.

### Data Quality Discussion
- **Positive criterion (Task relevance):** Includes direct exposure and engagement indicators (`views`, `likes`, `comments`) that are central to reel virality analysis.
- **Problematic criterion (Data validity/completeness):** Presence of invalid negatives and/or missing follower information can distort normalized metrics such as engagement rate if not handled correctly.

### Concluding Suitability Statement
This dataset is **suitable and important** for video-focused virality analysis, provided that invalid values are corrected to missing, normalisation is done carefully, and missing follower/view cases are handled transparently.

## **DS4 - Instagram Datasets**

In [1410]:
df_comments = pd.read_csv("datasets/raw/Instagram_Datasets.csv")

# Structural profile
print(df_comments.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   url                      1000 non-null   object
 1   comment_user             1000 non-null   object
 2   comment_user_url         1000 non-null   object
 3   comment_date             1000 non-null   object
 4   comment                  1000 non-null   object
 5   likes_number             1000 non-null   int64 
 6   replies_number           1000 non-null   int64 
 7   replies                  998 non-null    object
 8   hashtag_comment          6 non-null      object
 9   tagged_users_in_comment  146 non-null    object
 10  post_url                 1000 non-null   object
 11  post_user                1000 non-null   object
 12  comment_id               1000 non-null   int64 
 13  post_id                  1000 non-null   int64 
dtypes: int64(4), object(10)
memory usage: 109

`Instagram_Datasets.csv` is primarily **comment-level** data (`comment`, `comment_user`, `replies_number`, `tagged_users_in_comment`), linked to posts via `post_id` and `post_url`. It is less aligned with our main question because the analytical task focuses on **post characteristics vs viral status**, which is post-level rather than comment-thread level.

### Data Quality Discussion
- **Positive criterion (Textual depth):** Rich comment text and reply metadata can support sentiment studies or audience-reaction analysis.
- **Problematic criterion (Granularity mismatch):** Core unit is comments, not posts, so it does not directly provide post-level characteristics or viral labels needed for the current analytical objective.

### Concluding Suitability Statement
We will not move on with this dataset as it does not have post-level data which we need. It may be used as an optional enrichment source (e.g., aggregated sentiment per post), but only after substantial transformation and post-level linking.

## **DS5 - TikTok Posts**

In [1411]:
df_tiktok_posts = pd.read_csv("datasets/raw/TikTok_Posts.csv")

# Structural profile
print(df_tiktok_posts.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57960 entries, 0 to 57959
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   video_id         57960 non-null  int64  
 1   creator          57960 non-null  object 
 2   web_url          57960 non-null  object 
 3   date_posted      57960 non-null  object 
 4   duration_sec     57960 non-null  float64
 5   post_hour        57960 non-null  int64  
 6   post_day         57960 non-null  int64  
 7   is_ad            57960 non-null  bool   
 8   is_ai_generated  57960 non-null  bool   
dtypes: bool(2), float64(1), int64(3), object(3)
memory usage: 3.2+ MB
None


`TikTok_Posts.csv` is a post-level TikTok dataset containing core metadata (`video_id`, `creator`, `date_posted`, `post_hour`, `post_day`) and content/context indicators (`duration_sec`, `is_ad`, `is_ai_generated`). It is useful for modeling virality-related behavior by post timing and post attributes, but it does not include direct engagement outcomes (e.g., likes, comments, shares), so it should be treated as a supporting feature source rather than a standalone primary dataset.

In [1412]:
# Missingness and duplicates
print("\nMissing values:\n", df_tiktok_posts.isna().sum().sort_values(ascending=False))
print("Duplicate video_id:", df_tiktok_posts["video_id"].duplicated().sum())


Missing values:
 video_id           0
creator            0
web_url            0
date_posted        0
duration_sec       0
post_hour          0
post_day           0
is_ad              0
is_ai_generated    0
dtype: int64
Duplicate video_id: 0


Data profiling for DS5 is expected to show very low missingness because this file is schema-compact and mostly system-generated metadata. If duplicate `video_id` records are minimal or absent, the dataset has strong row-level integrity for joins with DS6 (`video_id` as the key).

### Data Quality Discussion
- **Positive criterion (Coverage/consistency):** The dataset is large and structurally consistent, with standardized temporal and creator-level metadata useful for feature engineering.
- **Problematic criterion (Target incompleteness):** No direct engagement outcomes are present, so DS5 alone cannot label or evaluate virality.

### Concluding Suitability Statement

## **DS6 - TikTok Texts**

In [1413]:
df_tiktok_texts = pd.read_csv("datasets/raw/TikTok_Texts.csv")

# Structural profile
print(df_tiktok_texts.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57960 entries, 0 to 57959
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   video_id       57960 non-null  int64 
 1   description    57549 non-null  object
 2   hashtags       57960 non-null  object
 3   hashtag_count  57960 non-null  int64 
dtypes: int64(2), object(2)
memory usage: 1.8+ MB
None


`TikTok_Texts.csv` is a text-focused TikTok dataset that contains `description`, `hashtags`, and `hashtag_count` keyed by `video_id`. It is highly relevant for extracting language and hashtag-based post characteristics, but it does not provide direct engagement outcomes, so it functions best as a complementary feature layer when joined with post-level datasets.

In [1414]:
# Missingness and duplicates
print("\nMissing values:\n", df_tiktok_texts.isna().sum().sort_values(ascending=False))
print("Duplicate video_id:", df_tiktok_texts["video_id"].duplicated().sum())


Missing values:
 description      411
video_id           0
hashtags           0
hashtag_count      0
dtype: int64
Duplicate video_id: 0


DS6 is expected to have low missingness except for occasional null `description` entries. Validity of `video_id` uniqueness is important because this key is required for reliable one-to-one or one-to-many joins with DS5 and any engagement dataset.

### Data Quality Discussion
- **Positive criterion (Textual relevance):** DS6 directly supports text-centric predictors such as caption length, hashtag usage, and lexical features.
- **Problematic criterion (Outcome absence):** There are no engagement or virality labels, so DS6 cannot independently answer the research question.

### Concluding Suitability Statement

# **6. Data Preparation (Process)**

Before we begin with data preparation, we will check how `engagement_rate` and `performance_bucket_label` in DS1 are defined as mentioned earlier.

### Validation of Pre-Derived Variables (DS1)

`Instagram_Analytics.csv` already includes `engagement_rate` and `performance_bucket_label`. Instead of assuming these fields are correct, we first re-derive them from the other columns and compare against the provided values.

This validation step lets us verify whether the dataset's derivation logic is consistent with our assumptions, quantify agreement using match rates, and decide whether the provided derived columns can be trusted for later use.

In [1415]:
import numpy as np

### Engagement rate:

Candidate formula for engagement_rate:

**engagement_rate = (likes + comments + shares + saves) / impressions**

The engagement rate formula measures how effectively a post generates interaction relative to how often it is shown. We sum all of the engagement metrics first. Dividing this total by impressions normalises the metric by the number of times the content appears on users’ screens, allowing fair comparison across posts with different visibility levels. As a result, the engagement rate reflects the likelihood that viewers interact with the content each time it is displayed, making it a useful indicator of content effectiveness.

In [1416]:
# Re-derive engagement_rate
df_analytics["derived_engagement_rate"] = (
    (df_analytics["likes"] + df_analytics["comments"] + df_analytics["shares"] + df_analytics["saves"])
    / df_analytics["impressions"]
)

# Round to 3s.f.
df_analytics["derived_engagement_rate"] = df_analytics["derived_engagement_rate"].map(lambda x: float(f"{x:.3g}"))

# Compare with provided engagement_rate
engagement_close = np.isclose(
    df_analytics["derived_engagement_rate"],
    df_analytics["engagement_rate"],
    rtol=1e-5,
    atol=1e-8,
    equal_nan=True,
)

# Count of mismatches
mismatch_count = (engagement_close == False).sum()
print("Rows not close:", mismatch_count, f"out of {len(df_analytics)} rows")

# Percentage of mismatch
print(f"Mismatch rate: {(mismatch_count / len(df_analytics))*100:.2f}%")

Rows not close: 1217 out of 29999 rows
Mismatch rate: 4.06%


Since the mismatch rate between the `derived engagement rate` and the provided `engagement rate` is very low, we infer that the formula used to compute the `derived engagement rate` is likely consistent with the one used in the dataset. As this formulation also aligns well with our research question on measuring user interaction relative to content exposure, we can proceed to use the `engagement_rate` column provided in DS1 for further analysis.

In [1417]:
# Drop created column after use
df_analytics = df_analytics.drop(columns=["derived_engagement_rate"])

### Performance bucket label:

Candidate formula for performance_bucket_label:

Based on the distribution of `df_analytics["engagement_rate"]`, we split it into 4 quartiles to obtain the labels `low`, `medium`, `high` and `viral`.

In [1418]:
labels = df_analytics.groupby("performance_bucket_label")["performance_bucket_label"].count()
print(labels, '\n')

# Split the engagement rate into 4 quartiles for bucket labelling
derived_ranked = df_analytics["engagement_rate"].rank(method="first")
derived_bucket = pd.qcut(
    derived_ranked,
    q=4,
    labels=["low", "medium", "high", "viral"],
)

provided_bucket_label = df_analytics["performance_bucket_label"].astype(str)
derived_bucket_label = derived_bucket.astype(str)

bucket_match = provided_bucket_label.eq(derived_bucket_label)

# Count of mismatches
bucket_not_match_count = (bucket_match == False).sum()
print("Buckets not match:", bucket_not_match_count, f"out of {len(df_analytics)} rows")

# Percentage of mismatch
print(f"Mismatch rate: {(bucket_not_match_count / len(df_analytics))*100:.2f}%")

performance_bucket_label
high      7500
low       7500
medium    7499
viral     7500
Name: performance_bucket_label, dtype: int64 

Buckets not match: 3347 out of 29999 rows
Mismatch rate: 11.16%


Since the mismatch rate between the `derived engagement buckets` and the provided `performance_bucket_label` is relatively low (11.16%), the provided labels appear to be largely consistent with the quartile-based classification derived from the engagement rate. This indicates that the dataset’s existing bucket labels reasonably reflect the engagement distribution. Therefore, it is acceptable to use the `performance_bucket_label` column directly in subsequent analysis.

## Preparation Strategy

After loading and profiling the datasets, we proceed with a structured preparation workflow to ensure consistency and quality.

1. **Standardise structure** by aligning data types, datetime formats, and field naming across datasets; remove non-informative or irrelevant columns.
2. **Clean invalid values** (e.g., impossible negatives, malformed entries).
3. **Enrich with supporting attributes** (e.g., hashtag-based indicators, text-length features) that are meaningful for virality analysis.

## 1. Structuring of data

### DS1 Structuring

| Index | Column Name                | Action |
|-------|----------------------------|--------|
| 00 | post_id | **Keep** |
| 01 | account_id | Drop |
| 02 | account_type | Drop |
| 03 | follower_count | **Keep** |
| 04 | media_type | **Keep** |
| 05 | content_category | **Keep** |
| 06 | traffic_source | Drop |
| 07 | has_call_to_action | **Keep** |
| 08 | post_datetime | Drop |
| 09 | post_date | **Keep** |
| 10 | post_hour | **Keep** |
| 11 | day_of_week | **Keep** |
| 12 | likes | **Keep** |
| 13 | comments | **Keep** |
| 14 | shares | **Keep** |
| 15 | saves | **Keep** |
| 16 | reach | **Keep** |
| 17 | impressions | **Keep** |
| 18 | engagement_rate | **Keep** |
| 19 | followers_gained | Drop |
| 20 | caption_length | **Keep** |
| 21 | hashtags_count | **Keep** |
| 22 | performance_bucket_label | **Keep** |

In [1419]:
# Check contents
print(df_analytics['account_type'].unique())
print(df_analytics['has_call_to_action'].unique())
print(df_analytics['post_datetime'].head(1))
print(df_analytics['post_date'].head(1))
print(df_analytics['post_hour'].head(1))

['brand' 'creator']
[1 0]
0    2024-11-30 06:00:00
Name: post_datetime, dtype: object
0    2024-11-30
Name: post_date, dtype: object
0    6
Name: post_hour, dtype: int64


In [1420]:
# Drop columns
ds1_drop_cols = ["account_id", "account_type", "traffic_source", "post_datetime", "followers_gained"]
df_ds1 = df_analytics.drop(columns=ds1_drop_cols).copy()

print(df_ds1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29999 entries, 0 to 29998
Data columns (total 18 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   post_id                   29999 non-null  object 
 1   follower_count            29999 non-null  int64  
 2   media_type                29999 non-null  object 
 3   content_category          29999 non-null  object 
 4   has_call_to_action        29999 non-null  int64  
 5   post_date                 29999 non-null  object 
 6   post_hour                 29999 non-null  int64  
 7   day_of_week               29999 non-null  object 
 8   likes                     29999 non-null  int64  
 9   comments                  29999 non-null  int64  
 10  shares                    29999 non-null  int64  
 11  saves                     29999 non-null  int64  
 12  reach                     29999 non-null  int64  
 13  impressions               29999 non-null  int64  
 14  engage

In [1421]:
# Rename columns
ds1_rename_map = {
    "media_type": "content_type",
    "day_of_week": "post_day_of_week",
    "comments": "comment_count",
    "likes": "like_count",
    "shares": "share_count",
    "saves": "save_count",
    "reach": "reach_count",
    "impressions": "impression_count",
    "hashtags_count": "hashtag_count"
}
df_ds1 = df_ds1.rename(columns=ds1_rename_map)

# Split post_date into separate calendar components
post_date_parsed = pd.to_datetime(df_ds1["post_date"])
df_ds1["post_day"] = post_date_parsed.dt.day.astype("Int64")
df_ds1["post_month"] = post_date_parsed.dt.month.astype("Int64")
df_ds1["post_year"] = post_date_parsed.dt.year.astype("Int64")

# Verification
print(df_ds1["post_date"].head(1))
print(df_ds1["post_day"].head(1))
print(df_ds1["post_month"].head(1))
print(df_ds1["post_year"].head(1))

df_ds1 = df_ds1.drop(columns=["post_date"])

0    2024-11-30
Name: post_date, dtype: object
0    30
Name: post_day, dtype: Int64
0    11
Name: post_month, dtype: Int64
0    2024
Name: post_year, dtype: Int64


In [1422]:
print(df_ds1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29999 entries, 0 to 29998
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   post_id                   29999 non-null  object 
 1   follower_count            29999 non-null  int64  
 2   content_type              29999 non-null  object 
 3   content_category          29999 non-null  object 
 4   has_call_to_action        29999 non-null  int64  
 5   post_hour                 29999 non-null  int64  
 6   post_day_of_week          29999 non-null  object 
 7   like_count                29999 non-null  int64  
 8   comment_count             29999 non-null  int64  
 9   share_count               29999 non-null  int64  
 10  save_count                29999 non-null  int64  
 11  reach_count               29999 non-null  int64  
 12  impression_count          29999 non-null  int64  
 13  engagement_rate           29999 non-null  float64
 14  captio

In [1423]:
# Datatype alignment

# String align
ds1_str_cols = ["post_id"]
for col in ds1_str_cols:
	df_ds1[col] = df_ds1[col].astype("string")

# Category align
ds1_cat_cols = ["content_type", "content_category", "post_day_of_week", "performance_bucket_label"]
for col in ds1_cat_cols:
	df_ds1[col] = df_ds1[col].astype("category")

# Integer align
ds1_int_cols = [
    "follower_count", "post_hour", "comment_count", "like_count", "share_count", "save_count", 
    "reach_count", "impression_count", "caption_length", "hashtag_count"
]
for col in ds1_int_cols:
    df_ds1[col] = pd.to_numeric(df_ds1[col]).astype("Int64")

# Float align
df_ds1["engagement_rate"] = pd.to_numeric(df_ds1["engagement_rate"]).astype("Float64")

# Boolean align
df_ds1["has_call_to_action"] = df_ds1["has_call_to_action"].astype("boolean")

# Dataset representation
df_ds1["dataset"] = "DS1"
df_ds1["dataset"] = df_ds1["dataset"].astype("category")

# Reorder columns
post_detail_cols = [
    "content_type", "content_category", "caption_length", "post_day_of_week",
    "post_hour", "post_day", "post_month", "post_year"
]
independent_cols = ["follower_count", "hashtag_count", "has_call_to_action"]
engagement_cols = [
    "like_count", "comment_count", "share_count", "save_count",
    "reach_count", "impression_count"
]
derived_cols = ["engagement_rate", "performance_bucket_label", "dataset"]

ordered_cols = ["post_id"]
ordered_cols += [col for col in post_detail_cols]
ordered_cols += [col for col in independent_cols]
ordered_cols += [col for col in engagement_cols]
ordered_cols += [col for col in derived_cols]

df_ds1 = df_ds1[ordered_cols]

print(df_ds1.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29999 entries, 0 to 29998
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype   
---  ------                    --------------  -----   
 0   post_id                   29999 non-null  string  
 1   content_type              29999 non-null  category
 2   content_category          29999 non-null  category
 3   caption_length            29999 non-null  Int64   
 4   post_day_of_week          29999 non-null  category
 5   post_hour                 29999 non-null  Int64   
 6   post_day                  29999 non-null  Int64   
 7   post_month                29999 non-null  Int64   
 8   post_year                 29999 non-null  Int64   
 9   follower_count            29999 non-null  Int64   
 10  hashtag_count             29999 non-null  Int64   
 11  has_call_to_action        29999 non-null  boolean 
 12  like_count                29999 non-null  Int64   
 13  comment_count             29999 non-null  Int6

### DS2 Structuring

| Index | Column Name              | Action |
|-------|--------------------------|--------|
| 00 | url | Drop |
| 01 | user_posted | Drop |
| 02 | description | **Keep** |
| 03 | hashtags | **Keep** |
| 04 | num_comments | **Keep** |
| 05 | date_posted | **Keep** |
| 06 | likes | **Keep** |
| 07 | photos | Drop |
| 08 | videos | Drop |
| 09 | location | Drop |
| 10 | latest_comments | Drop |
| 11 | post_id | **Keep** |
| 12 | discovery_input | Drop |
| 13 | has_handshake | Drop |
| 14 | shortcode | Drop |
| 15 | content_type | **Keep** |
| 16 | pk | Drop |
| 17 | content_id | Drop |
| 18 | engagement_score_view | Drop |
| 19 | thumbnail | Drop |
| 20 | video_view_count | Drop |
| 21 | product_type | Drop |
| 22 | coauthor_producers | Drop |
| 23 | tagged_users | **Keep** |
| 24 | video_play_count | Drop |
| 25 | followers | **Keep** |
| 26 | posts_count | Drop |
| 27 | profile_image_link | Drop |
| 28 | is_verified | **Keep** |
| 29 | is_paid_partnership | Drop |
| 30 | partnership_details | Drop |
| 31 | user_posted_id | Drop |
| 32 | post_content | Drop |
| 33 | audio | Drop |
| 34 | profile_url | Drop |
| 35 | videos_duration | Drop |
| 36 | images | Drop |
| 37 | alt_text | Drop |
| 38 | photos_number | Drop |
| 39 | audio_url | Drop |

In [1424]:
# Check contents
print(df_posts["user_posted"].head(1))
print(df_posts["photos"].head(1))
print(df_posts["videos"].head(1))
print(df_posts["pk"].head(1))
print(df_posts["product_type"].unique())
print(df_posts["coauthor_producers"].head(1))
print(df_posts["tagged_users"].head(1))
print(df_posts["is_verified"].unique())
print(df_posts["date_posted"].head(1))
print(df_posts["post_content"].head(1))
print(df_posts["alt_text"].head(1))

0    joyrichhhhh
Name: user_posted, dtype: object
0    ["https://scontent-dus1-1.cdninstagram.com/v/t...
Name: photos, dtype: object
0    NaN
Name: videos, dtype: object
0    3554806872009530025
Name: pk, dtype: int64
[nan 'clips' 'igtv' 'feed']
0    NaN
Name: coauthor_producers, dtype: object
0    NaN
Name: tagged_users, dtype: object
[False  True]
0    2025-01-27T13:53:43.000Z
Name: date_posted, dtype: object
0    [{"alt_text":"Photo by เสื้อผ้าสายฝอ ไปทะเล พร...
Name: post_content, dtype: object
0    Photo by เสื้อผ้าสายฝอ ไปทะเล พร้อมส่ง on Janu...
Name: alt_text, dtype: object


In [1425]:
# Drop columns
ds2_drop_cols = [
    "url", "user_posted", "photos", "videos", "location", "latest_comments",
    "discovery_input", "has_handshake", "shortcode", "pk", "content_id", "engagement_score_view",
    "thumbnail", "video_view_count", "product_type", "coauthor_producers", "tagged_users",
    "video_play_count", "posts_count", "profile_image_link", "partnership_details",
    "user_posted_id", "post_content", "audio", "profile_url", "videos_duration", "images",
    "alt_text", "photos_number", "audio_url"
]
df_ds2 = df_posts.drop(columns=ds2_drop_cols).copy()

print(df_ds2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   description          956 non-null    object 
 1   hashtags             647 non-null    object 
 2   num_comments         1000 non-null   int64  
 3   date_posted          1000 non-null   object 
 4   likes                888 non-null    float64
 5   post_id              1000 non-null   int64  
 6   content_type         1000 non-null   object 
 7   followers            1000 non-null   int64  
 8   is_verified          1000 non-null   bool   
 9   is_paid_partnership  1000 non-null   bool   
dtypes: bool(2), float64(1), int64(3), object(4)
memory usage: 64.6+ KB
None


In [1426]:
# Rename columns
ds2_rename_map = {
    "description": "caption",
    "num_comments": "comment_count",
    "date_posted": "post_date",
    "likes": "like_count",
    "followers": "follower_count"
}
df_ds2 = df_ds2.rename(columns=ds2_rename_map)

# Split post_date into separate calendar components
post_date_parsed = pd.to_datetime(df_ds2["post_date"])
df_ds2["post_day"] = post_date_parsed.dt.day.astype("Int64")
df_ds2["post_month"] = post_date_parsed.dt.month.astype("Int64")
df_ds2["post_year"] = post_date_parsed.dt.year.astype("Int64")
df_ds2["post_hour"] = post_date_parsed.dt.hour.astype("Int64")
df_ds2["post_day_of_week"] = post_date_parsed.dt.day_name().astype("category")

# Derive supporting fields
df_ds2["caption_length"] = df_ds2["caption"].fillna("").astype(str).str.len()
df_ds2["hashtag_count"] = df_ds2["hashtags"].fillna("").astype(str).str.count(r"#[A-Za-z0-9_]+")
df_ds2["engagement_rate"] = (df_ds2["like_count"] + df_ds2["comment_count"]) / (df_ds2["follower_count"] + 1)

# Verification
print(df_ds2["post_date"].head(1))
print(df_ds2["post_day"].head(1))
print(df_ds2["post_month"].head(1))
print(df_ds2["post_year"].head(1))
print(df_ds2["post_day_of_week"].head(1))

df_ds2 = df_ds2.drop(columns=["post_date", "caption"])

0    2025-01-27T13:53:43.000Z
Name: post_date, dtype: object
0    27
Name: post_day, dtype: Int64
0    1
Name: post_month, dtype: Int64
0    2025
Name: post_year, dtype: Int64
0    Monday
Name: post_day_of_week, dtype: category
Categories (7, object): ['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']


In [1427]:
print(df_ds2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   hashtags             647 non-null    object  
 1   comment_count        1000 non-null   int64   
 2   like_count           888 non-null    float64 
 3   post_id              1000 non-null   int64   
 4   content_type         1000 non-null   object  
 5   follower_count       1000 non-null   int64   
 6   is_verified          1000 non-null   bool    
 7   is_paid_partnership  1000 non-null   bool    
 8   post_day             1000 non-null   Int64   
 9   post_month           1000 non-null   Int64   
 10  post_year            1000 non-null   Int64   
 11  post_hour            1000 non-null   Int64   
 12  post_day_of_week     1000 non-null   category
 13  caption_length       1000 non-null   int64   
 14  hashtag_count        1000 non-null   int64   
 15  engagement_rate      8

In [1428]:
# Datatype alignment

# String align
ds2_str_cols = ["post_id"]
for col in ds2_str_cols:
    df_ds2[col] = df_ds2[col].astype("string")

# Category align
ds2_cat_cols = ["content_type"]
for col in ds2_cat_cols:
    df_ds2[col] = df_ds2[col].astype("category")

# Integer align
ds2_int_cols = [
    "comment_count", "like_count",
    "follower_count", "caption_length", "hashtag_count"
]
for col in ds2_int_cols:
    df_ds2[col] = pd.to_numeric(df_ds2[col]).astype("Int64")

# Float align
df_ds2["engagement_rate"] = pd.to_numeric(df_ds2["engagement_rate"]).astype("Float64")

# Boolean align
df_ds2["is_verified"] = df_ds2["is_verified"].astype("boolean")
df_ds2["is_paid_partnership"] = df_ds2["is_paid_partnership"].astype("boolean")

# Dataset representation
df_ds2["dataset"] = "DS2"
df_ds2["dataset"] = df_ds2["dataset"].astype("category")

# Reorder columns
post_detail_cols = [
    "content_type", "caption_length", "post_day_of_week",
    "post_hour", "post_day", "post_month", "post_year"
]
independent_cols = ["follower_count", "hashtag_count", "is_verified", "is_paid_partnership", "hashtags"]
engagement_cols = ["like_count", "comment_count"]
derived_cols = ["engagement_rate", "dataset"]

ordered_cols = ["post_id"]
ordered_cols += [col for col in post_detail_cols]
ordered_cols += [col for col in independent_cols]
ordered_cols += [col for col in engagement_cols]
ordered_cols += [col for col in derived_cols]

df_ds2 = df_ds2[ordered_cols]

print(df_ds2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   post_id              1000 non-null   string  
 1   content_type         1000 non-null   category
 2   caption_length       1000 non-null   Int64   
 3   post_day_of_week     1000 non-null   category
 4   post_hour            1000 non-null   Int64   
 5   post_day             1000 non-null   Int64   
 6   post_month           1000 non-null   Int64   
 7   post_year            1000 non-null   Int64   
 8   follower_count       1000 non-null   Int64   
 9   hashtag_count        1000 non-null   Int64   
 10  is_verified          1000 non-null   boolean 
 11  is_paid_partnership  1000 non-null   boolean 
 12  hashtags             647 non-null    object  
 13  like_count           888 non-null    Int64   
 14  comment_count        1000 non-null   Int64   
 15  engagement_rate      8

### DS3 Structuring

| Index | Column Name           | Action |
|-------|-----------------------|--------|
| 0  | url | Drop |
| 1  | user_posted | Drop |
| 2  | description | **Keep** |
| 3  | hashtags | **Keep** |
| 4  | num_comments | **Keep** |
| 5  | date_posted | **Keep** |
| 6  | likes | **Keep** |
| 7  | views | **Keep** |
| 8  | video_play_count | **Keep** |
| 9  | top_comments | Drop |
| 10 | post_id | **Keep** |
| 11 | thumbnail | Drop |
| 12 | shortcode | Drop |
| 13 | content_id | Drop |
| 14 | product_type | Drop |
| 15 | coauthor_producers | **Keep** |
| 16 | tagged_users | **Keep** |
| 17 | length | **Keep** |
| 18 | video_url | Drop |
| 19 | audio_url | Drop |
| 20 | posts_count | Drop |
| 21 | followers | **Keep** |
| 22 | following | Drop |
| 23 | user_profile_url | Drop |
| 24 | is_paid_partnership | **Keep** |
| 25 | is_verified | **Keep** |
| 26 | profile_image_link | Drop |

In [1429]:
# Check contents
print(df_reels["product_type"].unique())
print(df_reels["coauthor_producers"].head(5))
print(df_reels["tagged_users"].head(5))

['clips' 'feed' 'igtv']
0                                                   []
1    ["officeofaut","yuvasenaforyou","shivsenaubtco...
2                                                   []
3                                                   []
4                               ["indonesiajuaratrip"]
Name: coauthor_producers, dtype: object
0                                                   []
1    [{"full_name":"Aditya Thackeray","id":"1787743...
2                                                   []
3    [{"full_name":"Christopher Villa","id":"906480...
4    [{"full_name":"OPEN TRIP LABUAN BAJO | KOMODO ...
Name: tagged_users, dtype: object


In [1430]:
# Drop columns
ds3_drop_cols = [
    "url", "user_posted", "top_comments", "thumbnail", "shortcode", "content_id","product_type",
    "video_url", "audio_url", "posts_count", "following", "user_profile_url", "profile_image_link"
]
df_ds3 = df_reels.drop(columns=ds3_drop_cols).copy()

print(df_ds3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   description          956 non-null    object 
 1   hashtags             656 non-null    object 
 2   num_comments         1000 non-null   int64  
 3   date_posted          1000 non-null   object 
 4   likes                1000 non-null   int64  
 5   views                1000 non-null   int64  
 6   video_play_count     984 non-null    float64
 7   post_id              1000 non-null   object 
 8   coauthor_producers   1000 non-null   object 
 9   tagged_users         1000 non-null   object 
 10  length               1000 non-null   float64
 11  followers            519 non-null    float64
 12  is_paid_partnership  1000 non-null   bool   
 13  is_verified          1000 non-null   bool   
dtypes: bool(2), float64(3), int64(3), object(6)
memory usage: 95.8+ KB
None


In [1431]:
# Rename columns
ds3_rename_map = {
    "description": "caption",
    "num_comments": "comment_count",
    "date_posted": "post_date",
    "likes": "like_count",
    "views": "view_count",
    "video_play_count": "play_count",
    "length": "video_duration_sec",
    "followers": "follower_count",
}
df_ds3 = df_ds3.rename(columns=ds3_rename_map)

# Split post_date into separate calendar components
post_date_parsed = pd.to_datetime(df_ds3["post_date"])
df_ds3["post_day"] = post_date_parsed.dt.day.astype("Int64")
df_ds3["post_month"] = post_date_parsed.dt.month.astype("Int64")
df_ds3["post_year"] = post_date_parsed.dt.year.astype("Int64")
df_ds3["post_hour"] = post_date_parsed.dt.hour.astype("Int64")
df_ds3["post_day_of_week"] = post_date_parsed.dt.day_name().astype("category")

# Derive supporting fields
df_ds3["content_type"] = "reel"
df_ds3["caption_length"] = df_ds3["caption"].fillna("").astype(str).str.len()
df_ds3["hashtag_count"] = df_ds3["hashtags"].fillna("").astype(str).str.count(r"#[A-Za-z0-9_]+")
df_ds3["engagement_rate"] = (df_ds3["like_count"] + df_ds3["comment_count"]) / (df_ds3["follower_count"] + 1)

# Verification
print(df_ds3["post_date"].head(1))
print(df_ds3["post_day"].head(1))
print(df_ds3["post_month"].head(1))
print(df_ds3["post_year"].head(1))
print(df_ds3["post_day_of_week"].head(1))

df_ds3 = df_ds3.drop(columns=["post_date", "caption"])

0    2025-01-24T09:11:16.000Z
Name: post_date, dtype: object
0    24
Name: post_day, dtype: Int64
0    1
Name: post_month, dtype: Int64
0    2025
Name: post_year, dtype: Int64
0    Friday
Name: post_day_of_week, dtype: category
Categories (7, object): ['Friday', 'Monday', 'Saturday', 'Sunday', 'Thursday', 'Tuesday', 'Wednesday']


In [1432]:
print(df_ds3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   hashtags             656 non-null    object  
 1   comment_count        1000 non-null   int64   
 2   like_count           1000 non-null   int64   
 3   view_count           1000 non-null   int64   
 4   play_count           984 non-null    float64 
 5   post_id              1000 non-null   object  
 6   coauthor_producers   1000 non-null   object  
 7   tagged_users         1000 non-null   object  
 8   video_duration_sec   1000 non-null   float64 
 9   follower_count       519 non-null    float64 
 10  is_paid_partnership  1000 non-null   bool    
 11  is_verified          1000 non-null   bool    
 12  post_day             1000 non-null   Int64   
 13  post_month           1000 non-null   Int64   
 14  post_year            1000 non-null   Int64   
 15  post_hour            1

In [1433]:
# Datatype alignment

# String align
ds3_str_cols = ["post_id"]
for col in ds3_str_cols:
    df_ds3[col] = df_ds3[col].astype("string")

# Category align
ds3_cat_cols = ["content_type"]
for col in ds3_cat_cols:
    df_ds3[col] = df_ds3[col].astype("category")

# Integer align
ds3_int_cols = [
    "follower_count", "comment_count", "like_count", "view_count",
    "play_count", "caption_length", "hashtag_count"
]
for col in ds3_int_cols:
    df_ds3[col] = pd.to_numeric(df_ds3[col]).astype("Int64")

# Float align
df_ds3["engagement_rate"] = pd.to_numeric(df_ds3["engagement_rate"]).astype("Float64")
df_ds3["video_duration_sec"] = pd.to_numeric(df_ds3["video_duration_sec"]).astype("Float64")

# Boolean align
df_ds3["is_verified"] = df_ds3["is_verified"].astype("boolean")
df_ds3["is_paid_partnership"] = df_ds3["is_paid_partnership"].astype("boolean")

# Dataset representation
df_ds3["dataset"] = "DS3"
df_ds3["dataset"] = df_ds3["dataset"].astype("category")

# Reorder columns
post_detail_cols = [
    "content_type", "video_duration_sec", "caption_length", "post_day_of_week",
    "post_hour", "post_day", "post_month", "post_year"
]
independent_cols = ["follower_count", "hashtag_count", "is_verified", "is_paid_partnership", "hashtags"]
engagement_cols = ["like_count", "comment_count", "view_count", "play_count"]
derived_cols = ["engagement_rate", "dataset"]

ordered_cols = ["post_id"]
ordered_cols += [col for col in post_detail_cols]
ordered_cols += [col for col in independent_cols]
ordered_cols += [col for col in engagement_cols]
ordered_cols += [col for col in derived_cols]

df_ds3 = df_ds3[ordered_cols]

print(df_ds3.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   post_id              1000 non-null   string  
 1   content_type         1000 non-null   category
 2   video_duration_sec   1000 non-null   Float64 
 3   caption_length       1000 non-null   Int64   
 4   post_day_of_week     1000 non-null   category
 5   post_hour            1000 non-null   Int64   
 6   post_day             1000 non-null   Int64   
 7   post_month           1000 non-null   Int64   
 8   post_year            1000 non-null   Int64   
 9   follower_count       519 non-null    Int64   
 10  hashtag_count        1000 non-null   Int64   
 11  is_verified          1000 non-null   boolean 
 12  is_paid_partnership  1000 non-null   boolean 
 13  hashtags             656 non-null    object  
 14  like_count           1000 non-null   Int64   
 15  comment_count        1

## 2. Cleaning of data

### Invalid-value handling and median imputation

In this step, we perform two actions across `df_ds1`, `df_ds2`, and `df_ds3`:

1. Convert invalid negative values in non-negative numeric metrics to `NaN`.
2. Impute all remaining numeric missing values using each column's `median`.

No actions needed for `df_ds1`, as there are no negative values and no null values in the dataset as checked earlier.

In [1434]:
def clean_non_negative(df, cols):
    """Helper function to convert negative values to NaN given a dataframe's columns."""
    df_copy = df.copy()

    # Convert
    negative_counts = {}
    for col in cols:
        mask = df_copy[col] < 0
        negative_counts[col] = int(mask.sum())
        df_copy.loc[mask, col] = pd.NA

    print("Negatives converted to NA:", {k: v for k, v in negative_counts.items() if v > 0} or "None")

    return df_copy

`df_ds2`

In [1435]:
# Null values:
# 12  hashtags             647 non-null    object  
# 13  like_count           888 non-null    Int64    
# 15  engagement_rate      888 non-null    Float64 

# Non-negative numerical fields for DS2
ds2_non_negative_numeric_cols = [
    "comment_count", "like_count", "follower_count",
    "post_day", "post_month", "post_year", "post_hour",
    "caption_length", "hashtag_count", "engagement_rate"
]

df_ds2 = clean_non_negative(df_ds2, ds2_non_negative_numeric_cols)

# hashtags -> empty list
df_ds2["hashtags"] = df_ds2["hashtags"].apply(lambda x: [] if pd.isna(x) else x)

# like_count -> median by content_type, then global median fallback
df_ds2["like_count"] = pd.to_numeric(df_ds2["like_count"])
group_median = df_ds2.groupby("content_type")["like_count"].transform("median")
global_median = df_ds2["like_count"].median()
df_ds2["like_count"] = df_ds2["like_count"].fillna(group_median).fillna(global_median)
df_ds2["like_count"] = df_ds2["like_count"].round().astype("Int64")

# engagement_rate -> recompute
df_ds2["engagement_rate"] = ((df_ds2["like_count"] + df_ds2["comment_count"]) / (df_ds2["follower_count"] + 1)).astype("Float64")

print(df_ds2.info())

Negatives converted to NA: None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   post_id              1000 non-null   string  
 1   content_type         1000 non-null   category
 2   caption_length       1000 non-null   Int64   
 3   post_day_of_week     1000 non-null   category
 4   post_hour            1000 non-null   Int64   
 5   post_day             1000 non-null   Int64   
 6   post_month           1000 non-null   Int64   
 7   post_year            1000 non-null   Int64   
 8   follower_count       1000 non-null   Int64   
 9   hashtag_count        1000 non-null   Int64   
 10  is_verified          1000 non-null   boolean 
 11  is_paid_partnership  1000 non-null   boolean 
 12  hashtags             1000 non-null   object  
 13  like_count           1000 non-null   Int64   
 14  comment_count        1000 non-null   Int6

C:\Users\User\AppData\Local\Temp\ipykernel_30764\1677613855.py:20: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group_median = df_ds2.groupby("content_type")["like_count"].transform("median")


`df_ds3`

In [1436]:
# Null values:
# 9   follower_count       519 non-null    Int64   
# 13  hashtags             656 non-null    object  
# 17  play_count           984 non-null    Int64   
# 18  engagement_rate      519 non-null    Float64 

# Non-negative numerical fields for DS3
ds3_non_negative_numeric_cols = [
    "comment_count", "like_count", "view_count", "play_count",
    "video_duration_sec", "follower_count",
    "post_day", "post_month", "post_year", "post_hour",
    "caption_length", "hashtag_count", "engagement_rate"
]

df_ds3 = clean_non_negative(df_ds3, ds3_non_negative_numeric_cols)

# hashtags -> empty list
df_ds3["hashtags"] = df_ds3["hashtags"].apply(lambda x: [] if pd.isna(x) else x)

# follower_count -> median by is_verified, fallback to global median
df_ds3["follower_count"] = pd.to_numeric(df_ds3["follower_count"], errors="coerce")
follower_group_median = df_ds3.groupby("is_verified")["follower_count"].transform("median")
follower_global_median = df_ds3["follower_count"].median()
df_ds3["follower_count"] = (
    df_ds3["follower_count"]
    .fillna(follower_group_median)
    .fillna(follower_global_median)
    .round()
    .astype("Int64")
)

# play_count -> median by content_type, fallback to global median
df_ds3["play_count"] = pd.to_numeric(df_ds3["play_count"], errors="coerce")
play_group_median = df_ds3.groupby("content_type")["play_count"].transform("median")
play_global_median = df_ds3["play_count"].median()
df_ds3["play_count"] = (
    df_ds3["play_count"]
    .fillna(play_group_median)
    .fillna(play_global_median)
    .round()
    .astype("Int64")
)

# engagement_rate -> recompute from counts and followers (instead of direct imputation)
df_ds3["engagement_rate"] = (
    (df_ds3["like_count"] + df_ds3["comment_count"]) / (df_ds3["follower_count"] + 1)
).astype("Float64")

print(df_ds3.info())

Negatives converted to NA: {'like_count': 57}
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype   
---  ------               --------------  -----   
 0   post_id              1000 non-null   string  
 1   content_type         1000 non-null   category
 2   video_duration_sec   1000 non-null   Float64 
 3   caption_length       1000 non-null   Int64   
 4   post_day_of_week     1000 non-null   category
 5   post_hour            1000 non-null   Int64   
 6   post_day             1000 non-null   Int64   
 7   post_month           1000 non-null   Int64   
 8   post_year            1000 non-null   Int64   
 9   follower_count       1000 non-null   Int64   
 10  hashtag_count        1000 non-null   Int64   
 11  is_verified          1000 non-null   boolean 
 12  is_paid_partnership  1000 non-null   boolean 
 13  hashtags             1000 non-null   object  
 14  like_count           943 no

C:\Users\User\AppData\Local\Temp\ipykernel_30764\3271456296.py:34: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  play_group_median = df_ds3.groupby("content_type")["play_count"].transform("median")


## 3. Enriching of data

# **7. Data Analysis (Analyze)**

### RQ1 — Hashtags & Discoverability

**Research question:** Does hashtag usage (presence and count) increase the probability of viral engagement?

We evaluate:
1. Whether posts with hashtags have higher engagement than posts without hashtags.
2. Whether higher hashtag counts are associated with higher engagement.
3. For DS2/DS3, whether hashtag *type mix* (broad vs niche) relates to engagement differences.

In [1437]:
import matplotlib.pyplot as plt
import seaborn as sns

try:
    from scipy import stats
    SCIPY_AVAILABLE = True
except Exception:
    SCIPY_AVAILABLE = False

sns.set_theme(style="whitegrid")

# Combine enriched datasets for unified analysis
analysis_df = pd.concat(
    [
        df_ds1.assign(dataset="DS1"),
        df_ds2.assign(dataset="DS2"),
        df_ds3.assign(dataset="DS3"),
    ],
    ignore_index=True,
)

# Keep rows with usable engagement metric and hashtag count
rq1_df = analysis_df.dropna(subset=["engagement_rate_for_rq1", "hashtag_count", "has_hashtag"]).copy()

# --- Descriptive summaries ---
print("RQ1 summary by dataset and hashtag presence")
rq1_presence = (
    rq1_df.groupby(["dataset", "has_hashtag"], as_index=False)
    .agg(
        posts=("engagement_rate_for_rq1", "size"),
        mean_engagement=("engagement_rate_for_rq1", "mean"),
        median_engagement=("engagement_rate_for_rq1", "median"),
        mean_hashtag_count=("hashtag_count", "mean"),
    )
)
print(rq1_presence)

print("\nSpearman correlation: hashtag_count vs engagement_rate_for_rq1")
for ds in ["DS1", "DS2", "DS3"]:
    s = rq1_df[rq1_df["dataset"] == ds]
    if len(s) < 3:
        print(f"{ds}: insufficient data")
        continue
    if SCIPY_AVAILABLE:
        corr, pval = stats.spearmanr(s["hashtag_count"], s["engagement_rate_for_rq1"], nan_policy="omit")
        print(f"{ds}: rho={corr:.4f}, p-value={pval:.4g}")
    else:
        corr = s[["hashtag_count", "engagement_rate_for_rq1"]].corr(method="spearman").iloc[0, 1]
        print(f"{ds}: rho={corr:.4f} (p-value unavailable; scipy not installed)")

# --- Statistical test for hashtag presence effect ---
print("\nHashtag presence effect test (has_hashtag=0 vs 1)")
for ds in ["DS1", "DS2", "DS3"]:
    s = rq1_df[rq1_df["dataset"] == ds]
    g0 = s.loc[s["has_hashtag"] == 0, "engagement_rate_for_rq1"]
    g1 = s.loc[s["has_hashtag"] == 1, "engagement_rate_for_rq1"]

    if len(g0) == 0 or len(g1) == 0:
        print(f"{ds}: skipped (only one hashtag-presence group found)")
        continue

    if SCIPY_AVAILABLE:
        u_stat, pval = stats.mannwhitneyu(g0, g1, alternative="two-sided")
        print(f"{ds}: Mann-Whitney U={u_stat:.2f}, p-value={pval:.4g}")
    else:
        print(f"{ds}: scipy unavailable, test skipped")

# --- DS2/DS3 refinement: broad vs niche mix ---
niche_df = rq1_df[rq1_df["dataset"].isin(["DS2", "DS3"])].copy()
if "parsed_hashtag_count" in niche_df.columns:
    def classify_niche(row):
        if row.get("parsed_hashtag_count", 0) == 0:
            return "no_hashtag"
        if row.get("niche_hashtag_ratio", 0.0) >= 0.7:
            return "niche_dominant"
        if row.get("broad_hashtag_ratio", 0.0) >= 0.7:
            return "broad_dominant"
        return "mixed"

    niche_df["hashtag_mix_type"] = niche_df.apply(classify_niche, axis=1)

    print("\nDS2/DS3 hashtag mix summary")
    print(
        niche_df.groupby(["dataset", "hashtag_mix_type"], as_index=False)
        .agg(
            posts=("engagement_rate_for_rq1", "size"),
            mean_engagement=("engagement_rate_for_rq1", "mean"),
            median_engagement=("engagement_rate_for_rq1", "median"),
        )
    )

# --- Visuals ---
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

sns.boxplot(
    data=rq1_df,
    x="dataset",
    y="engagement_rate_for_rq1",
    hue="has_hashtag",
    showfliers=False,
    ax=axes[0],
)
axes[0].set_title("Engagement by Hashtag Presence")
axes[0].set_xlabel("Dataset")
axes[0].set_ylabel("Engagement metric")
axes[0].legend(title="has_hashtag", loc="best")

scatter_df = rq1_df.copy()
# Downsample DS1 for readability while keeping DS2/DS3 full
mask_ds1 = scatter_df["dataset"] == "DS1"
scatter_df = pd.concat(
    [
        scatter_df[mask_ds1].sample(n=min(5000, mask_ds1.sum()), random_state=42),
        scatter_df[~mask_ds1],
    ],
    ignore_index=True,
)

sns.scatterplot(
    data=scatter_df,
    x="hashtag_count",
    y="engagement_rate_for_rq1",
    hue="dataset",
    alpha=0.35,
    s=20,
    ax=axes[1],
)
axes[1].set_title("Hashtag Count vs Engagement")
axes[1].set_xlabel("Hashtag count")
axes[1].set_ylabel("Engagement metric")

plt.tight_layout()
plt.show()

KeyError: ['engagement_rate_for_rq1', 'has_hashtag']

### RQ2 — Posting Time Effects

**Research question:** Does posting time (hour of day and weekday vs weekend) significantly affect the likelihood of a post going viral?

We analyze timing effects using:
1. `time_block` (`night`, `morning`, `afternoon`, `evening`)
2. `is_weekend` (weekday vs weekend)

Outcome variable remains `engagement_rate_for_rq1` for comparability across datasets.

In [ ]:
rq2_df = analysis_df.dropna(subset=["engagement_rate_for_rq1", "time_block", "is_weekend"]).copy()

# Keep expected categories only
valid_blocks = ["night", "morning", "afternoon", "evening"]
rq2_df = rq2_df[rq2_df["time_block"].isin(valid_blocks)].copy()
rq2_df["time_block"] = pd.Categorical(rq2_df["time_block"], categories=valid_blocks, ordered=True)
rq2_df["week_type"] = rq2_df["is_weekend"].map({0: "weekday", 1: "weekend"})

print("RQ2 summary by dataset and time_block")
summary_time = (
    rq2_df.groupby(["dataset", "time_block"], as_index=False)
    .agg(
        posts=("engagement_rate_for_rq1", "size"),
        mean_engagement=("engagement_rate_for_rq1", "mean"),
        median_engagement=("engagement_rate_for_rq1", "median"),
    )
)
print(summary_time)

print("\nRQ2 summary by dataset and week_type")
summary_week = (
    rq2_df.groupby(["dataset", "week_type"], as_index=False)
    .agg(
        posts=("engagement_rate_for_rq1", "size"),
        mean_engagement=("engagement_rate_for_rq1", "mean"),
        median_engagement=("engagement_rate_for_rq1", "median"),
    )
)
print(summary_week)

# --- Statistical tests ---
print("\nWeekend effect test (weekday vs weekend)")
for ds in ["DS1", "DS2", "DS3"]:
    s = rq2_df[rq2_df["dataset"] == ds]
    wd = s.loc[s["week_type"] == "weekday", "engagement_rate_for_rq1"]
    we = s.loc[s["week_type"] == "weekend", "engagement_rate_for_rq1"]

    if len(wd) == 0 or len(we) == 0:
        print(f"{ds}: skipped (missing weekday/weekend group)")
        continue

    if SCIPY_AVAILABLE:
        stat, pval = stats.mannwhitneyu(wd, we, alternative="two-sided")
        print(f"{ds}: Mann-Whitney U={stat:.2f}, p-value={pval:.4g}")
    else:
        print(f"{ds}: scipy unavailable, test skipped")

print("\nTime block effect test (night/morning/afternoon/evening)")
for ds in ["DS1", "DS2", "DS3"]:
    s = rq2_df[rq2_df["dataset"] == ds]
    groups = [
        s.loc[s["time_block"] == block, "engagement_rate_for_rq1"]
        for block in valid_blocks
    ]
    groups = [g for g in groups if len(g) > 0]

    if len(groups) < 2:
        print(f"{ds}: skipped (insufficient time-block groups)")
        continue

    if SCIPY_AVAILABLE:
        stat, pval = stats.kruskal(*groups)
        print(f"{ds}: Kruskal-Wallis H={stat:.2f}, p-value={pval:.4g}")
    else:
        print(f"{ds}: scipy unavailable, test skipped")

# --- Visuals ---
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

sns.barplot(
    data=summary_time,
    x="time_block",
    y="mean_engagement",
    hue="dataset",
    ax=axes[0],
)
axes[0].set_title("Mean Engagement by Time Block")
axes[0].set_xlabel("Time block")
axes[0].set_ylabel("Mean engagement")

sns.barplot(
    data=summary_week,
    x="week_type",
    y="mean_engagement",
    hue="dataset",
    ax=axes[1],
)
axes[1].set_title("Mean Engagement: Weekday vs Weekend")
axes[1].set_xlabel("Week type")
axes[1].set_ylabel("Mean engagement")

plt.tight_layout()
plt.show()

KeyError: ['time_block', 'is_weekend']